File: python/scripts/live_smoke.py
"""Real Qwen acceptance. No mock fallback; requires a locally configured API key."""

import argparse
import asyncio
from dataclasses import replace
from datetime import datetime, timezone
from pathlib import Path
import sys
import tempfile

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from agent.app import create_app
from agent.config import read_config
from agent.jsonutil import dumps, loads


def require(condition, message):
    if not condition:
        raise AssertionError(message)


def observations(session, tool_name):
    names, values = {}, []
    for message in session["messages"]:
        for call in message.get("tool_calls", []):
            names[call["id"]] = call["function"]["name"]
        if message["role"] == "tool" and names.get(message["tool_call_id"]) == tool_name:
            value = loads(message["content"])
            if value.get("ok"):
                values.append(value["data"])
    return values


async def run(report_path):
    # Never serialize Config: it contains the API credential.
    config = read_config()
    report = {"implementation": "python", "provider": config.provider, "model": config.model,
              "startedAt": datetime.now(timezone.utc).isoformat(), "passed": False, "turns": []}
    try:
        with tempfile.TemporaryDirectory(prefix="python-agent-live-") as data_dir:
            config = replace(config, data_dir=data_dir)
            app = create_app(config)
            a = await app.store.create("live-user", "天气和待办")
            b = await app.store.create("live-user", "周报和待办")

            async def turn(label, session, text, expected_tools=(), no_tools=False, instance=None):
                current = instance or app
                result = await current.runtime.run(user_id="live-user", session_id=session["id"], input=text)
                report["turns"].append({"label": label, "result": result})
                require(result["status"] == "ok", f"{label}: runtime did not finish successfully")
                calls = [e for e in result["trace"] if e["event"] == "tool.end"]
                for name in expected_tools:
                    require(any(e["tool"] == name and e["ok"] for e in calls), f"{label}: missing successful {name}")
                require(not no_tools or not calls, f"{label}: unexpected tool invocation")
                print(f"PASS {label} ({result['steps']} model calls, {len(calls)} tool executions)", flush=True)
                return result

            await turn("memory write", a, "我的项目代号是蓝鲸七号，请记住。", no_tools=True)
            result = await turn("conversation followup", a, "我的项目代号是什么？", no_tools=True)
            require("蓝鲸七号" in result["answer"], "conversation memory was not recalled")
            result = await turn("calculator", a, "请计算 (123+456)*7", ["calculator"])
            require("4053" in result["answer"], "incorrect calculator answer")
            require(observations(await app.store.get("live-user", a["id"]), "calculator")[-1]["result"] == 4053,
                    "incorrect calculator tool result")
            await turn("search", a, "请使用search搜索上下文压缩资料，并注明资料来源。", ["search"])
            search = observations(await app.store.get("live-user", a["id"]), "search")[-1]
            require(search["mock"] is True and search["source"] == "local-demo-corpus" and search["results"], "search source mismatch")
            await turn("weather and todo", a, "查询上海天气，并添加一条文字恰好为“明天带伞”的待办。", ["weather", "todo"])
            saved_a = await app.store.get("live-user", a["id"])
            weather = observations(saved_a, "weather")[-1]
            require(weather["mock"] is True and weather["city"] == "上海", "weather source mismatch")
            require(len(saved_a["todos"]) == 1 and saved_a["todos"][0]["text"] == "明天带伞", "first todo was not saved exactly")
            todo_id = saved_a["todos"][0]["id"]
            await turn("separate weekly report", b, "本周完成工具注册，下周补测试。写一段简短周报，并添加文字恰好为“周五提交周报”的待办。", ["todo"])
            saved_b = await app.store.get("live-user", b["id"])
            require(len(saved_b["todos"]) == 1 and saved_b["todos"][0]["text"] == "周五提交周报", "second session todo mismatch")
            await turn("tool followup", a, "把刚才明天带伞的待办标记完成。", ["todo"])
            saved_a = await app.store.get("live-user", a["id"])
            require(len(saved_a["todos"]) == 1 and saved_a["todos"][0]["id"] == todo_id and saved_a["todos"][0]["done"] is True,
                    "followup did not update the original todo")
            require(any(o.get("action") == "complete" for o in observations(saved_a, "todo")), "missing complete operation")
            restored = create_app(config)
            result = await turn("new runtime recovery", a, "我的项目代号是什么？明天带伞完成了吗？只查询，不要修改待办。", instance=restored)
            require("蓝鲸七号" in result["answer"] and "完成" in result["answer"], "restored context was not recalled")
            require((await restored.store.get("live-user", a["id"]))["todos"] == saved_a["todos"], "recovery changed first todos")
            require((await restored.store.get("live-user", b["id"]))["todos"] == saved_b["todos"], "session isolation failed")
            report["persistenceCheck"] = "passed: new Store/Runtime read same files; same process"
            report["passed"] = True
    except Exception as exc:
        # Only our synthetic assertion labels or a safe public error code are emitted.
        report["failure"] = str(exc) if isinstance(exc, AssertionError) else getattr(exc, "code", type(exc).__name__)
    finally:
        report["finishedAt"] = datetime.now(timezone.utc).isoformat()
        events = [e for turn in report["turns"] for e in turn["result"]["trace"]]
        report["modelCalls"] = sum(e["event"] == "llm.start" for e in events)
        report["toolExecutions"] = sum(e["event"] == "tool.end" for e in events)
        report["totalTokens"] = sum(e.get("usage", {}).get("total_tokens", 0) for e in events)
        report_path.parent.mkdir(parents=True, exist_ok=True)
        report_path.write_text(dumps(report) + "\n", encoding="utf-8")
        print(dumps({k: report[k] for k in ("passed", "modelCalls", "toolExecutions", "totalTokens")}))
        if not report["passed"]:
            print("FAIL " + report["failure"], file=sys.stderr)
    return 0 if report["passed"] else 1


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--report", type=Path,
                        default=Path(__file__).resolve().parents[2] / "docs/python-live-result.json")
    args = parser.parse_args()
    try:
        raise SystemExit(asyncio.run(run(args.report)))
    except Exception as exc:
        print("Live verification could not start: " + getattr(exc, "code", type(exc).__name__), file=sys.stderr)
        raise SystemExit(1)

File: javascript/scripts/live-smoke.js
import assert from 'node:assert/strict';
import { mkdtemp, mkdir, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { randomUUID } from 'node:crypto';
import { readConfig } from '../src/config.js';
import { createApp } from '../src/app.js';
import { SessionStore } from '../src/store.js';

const reportPath = fileURLToPath(new URL('../docs/live-result.json', import.meta.url));

function successfulToolData(result, tool) {
  return result.trace.filter(event => event.event === 'tool.end' && event.tool === tool && event.ok).map(event => {
    let observed;
    try { observed = JSON.parse(event.result); }
    catch { assert.fail(`${tool}: successful tool trace must include its complete bounded result JSON`); }
    assert.equal(observed.ok, true, `${tool}: tool trace must represent a successful observation`);
    return observed.data;
  });
}

function assertNoToolCalls(result, label) {
  assert.equal(result.trace.filter(event => event.event === 'tool.start').length, 0, `${label}: direct conversation must not call tools`);
}

// Explicit command only. Missing credentials cause failure, never a mocked success or skip.
const report = { startedAt: new Date().toISOString(), kind: 'real-llm-api', checks: [], passed: false };
try {
  const config = readConfig();
  config.dataDir = await mkdtemp(join(tmpdir(), 'minimal-agent-live-'));
  report.provider = config.provider; report.model = config.model;
  report.apiOrigin = new URL(config.baseUrl).origin;
  const { runtime, store } = createApp(config);
  const first = await store.create('live_A', '天气和待办');
  const second = await store.create('live_A', '周报和待办');
  const run = async (session, label, input, expectedTool, activeRuntime = runtime) => {
    console.log(`Running: ${label}`);
    const result = await activeRuntime.run({ userId: 'live_A', sessionId: session.id, input, requestId: randomUUID() });
    const check = { label, input, status: result.status, answer: result.answer, trace: result.trace };
    report.checks.push(check);
    assert.equal(result.status, 'ok', `${label}: ${result.answer}`);
    if (expectedTool) assert(result.trace.some(e => e.event === 'tool.end' && e.tool === expectedTool && e.ok), `${label}: expected successful ${expectedTool}`);
    return result;
  };
  const remembered = await run(first, '纯对话记忆写入', '请记住：我的项目代号是蓝鲸七号。简短确认即可，不需要工具。');
  assertNoToolCalls(remembered, '纯对话记忆写入');
  const follow = await run(first, '纯对话追问', '我的项目代号是什么？');
  assertNoToolCalls(follow, '纯对话追问');
  assert.match(follow.answer, /蓝鲸七号/);
  const calc = await run(first, '计算工具', '请用计算器算 (123+456)*7，并告诉我数值。', 'calculator');
  assert(successfulToolData(calc, 'calculator').some(data => data.result === 4053), 'calculator: observed result must be exactly 4053');
  assert.match(calc.answer, /4053|4,053/);
  const search = await run(first, '搜索工具', '请调用 search 搜索上下文压缩的资料，并说明数据来源。', 'search');
  assert(successfulToolData(search, 'search').some(data => data.mock === true && data.source === 'local-demo-corpus' && Array.isArray(data.results) && data.results.length > 0), 'search: expected nonempty results from the local mock corpus');
  assert.match(search.answer, /模拟|mock|演示|本地|local-demo-corpus/i, 'search: answer must disclose its mock/local source');
  const weather = await run(first, '查天气并写待办', '查询上海天气，并添加一条文字恰好为“明天带伞”的待办。', 'weather');
  assert(successfulToolData(weather, 'weather').some(data => data.city === '上海' && data.mock === true && data.source === 'fixed-demo-data'), 'weather: expected the Shanghai mock observation');
  assert(weather.trace.some(e => e.event === 'tool.end' && e.tool === 'todo' && e.ok));
  assert.match(weather.answer, /模拟|mock|演示|固定/i);
  await run(second, '独立窗口写周报和待办', '本周完成了 Agent 工具注册，下周准备补测试。帮我写一段简短周报，并添加一条文字恰好为“周五提交周报”的待办。', 'todo');
  const a = await store.get('live_A', first.id), b = await store.get('live_A', second.id);
  assert.deepEqual(a.todos.map(x => x.text), ['明天带伞']);
  assert.deepEqual(b.todos.map(x => x.text), ['周五提交周报']);
  await run(first, '带工具追问', '把刚才添加的带伞那项待办标记完成。', 'todo');
  const restarted = new SessionStore({ dataDir: config.dataDir });
  assert.equal((await restarted.get('live_A', first.id)).todos[0].done, true);
  assert.equal((await restarted.get('live_A', second.id)).todos[0].done, false);
  report.checks.push({ label: '重新读取持久化状态和双窗口隔离', status: 'ok' });
  // A fresh app creates a new runtime, client, context manager, and store over the saved files.
  const restartedApp = createApp(config);
  const restoredFirst = await restartedApp.store.get('live_A', first.id);
  const restoredSecond = await restartedApp.store.get('live_A', second.id);
  const afterRestart = await run(restoredFirst, '新 Runtime 续聊恢复记忆和待办', '我的项目代号是什么？“明天带伞”这项待办的状态是已完成还是未完成？请明确回答实际状态，只查询，不修改任何待办。', undefined, restartedApp.runtime);
  assert.match(afterRestart.answer, /蓝鲸七号/);
  assert.match(afterRestart.answer, /已.{0,8}完成|完成了|done\s*[:：=]\s*true/i, 'restart: answer must reflect the saved completed status');
  for (const observed of successfulToolData(afterRestart, 'todo')) {
    assert.equal(observed.action, 'list', 'restart: any successful todo invocation must be read-only');
  }
  assert.deepEqual((await restartedApp.store.get('live_A', first.id)).todos, restoredFirst.todos, 'restart: first-window todo state must remain unchanged');
  assert.deepEqual((await restartedApp.store.get('live_A', second.id)).todos, restoredSecond.todos, 'restart: second-window todo state must remain unchanged');
  report.passed = true;
  console.log('Real API smoke checks passed. See docs/live-result.json.');
} catch (err) {
  report.failure = { code: err.code || 'LIVE_CHECK_FAILED', message: String(err.message).slice(0, 600) };
  console.error(`Real API checks failed: ${report.failure.message}`);
  process.exitCode = 1;
} finally {
  report.finishedAt = new Date().toISOString();
  await mkdir(dirname(reportPath), { recursive: true });
  await writeFile(reportPath, JSON.stringify(report, null, 2) + '\n');
}
